content/assignments/minishell/util/parser/README.md

```diff
@@ -0,0 +1,59 @@
+# Parser
+
+The parser is made using [Bison](http://www.gnu.org/software/bison/) and [Flex](http://flex.sourceforge.net/).
+
+The parser is used to parse the commands entered by the user, stored in structure `command_t`.
+
+This structure is defined in the file `parser.h` and encapsulates a tree representation of the command.
+
+You can use `DisplayStructure.cpp` to display the structure using various [tests](### Tests).
+
+Also you can use `CUseParser.c` or `UseParser.cpp` to see how to use the parser in C or C++ and print the structure of the command.
+
+## Compile
+
+Run the following commands in the root of parser directory:
+
+```console
+student@os:/.../minishell/util/parser$ make
+```
+
+## Usage
+
+The parser is represented by two files:
+
+* `parser.y` - implementation of the parser
+* `parser.l` - implementation of the lexer
+
+### Build process
+
+Using Makefile, first generate the files `parser.yy.c` and `parser.tab.c` from `parser.y` and `parser.l`.
+After that, it compiles the files `parser.yy.c` and `parser.tab.c` to generate the object files `parser.yy.o` and `parser.tab.o`.
+To use the parser, you need to link the object files `parser.yy.o` and `parser.tab.o` with your program.
+
+### Example
+
+* `CUseParser.c` - example of using the parser in C
+* `UseParser.cpp` - example of using the parser in C++
+* `DisplayStructure.cpp` - reads multiple commands and displays the structure of the resulting tree
+
+### Tests
+
+More tests can be found in the `tests` directory:
+
+```console
+student@os:/.../minishell/util/parser$ cd tests
+student@os:/.../minishell/util/parser/tests$ ../DisplayStructure &>small_tests.out <small_tests.txt
+student@os:/.../minishell/util/parser/tests$ ../DisplayStructure &>ugly_tests.out <ugly_tests.txt
+student@os:/.../minishell/util/parser/tests$ ../DisplayStructure &>negative_tests.out <negative_tests.txt
```

Comment:
```suggestion
student@os:/.../minishell/util/parser$ cd tests

student@os:/.../minishell/util/parser/tests$ ../DisplayStructure &>small_tests.out <small_tests.txt

student@os:/.../minishell/util/parser/tests$ ../DisplayStructure &>ugly_tests.out <ugly_tests.txt

student@os:/.../minishell/util/parser/tests$ ../DisplayStructure &>negative_tests.out <negative_tests.txt
```
And add a few lines from the output so students know they're doing the right thing.